### PR TITLE
Add support for `samples`

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

With the latest development version of PennyLane, we now support returning `samples` from the QNode, as well as expectation values and variance:

```python
@qml.qnode(dev)
def circuit(x):
    qml.RX(x, wires=0)
    qml.CNOT(wires=[0, 1])
    return expval(qml.PauliZ(0)), sample(qml.PauliZ(1))
```

Plugins can support this new feature by subclassing the `Device.sample()` method, returning the requested number of samples for a given observable.

**Description of changes:**

* Adds support for returning samples from `wavefunction`, `numpy_wavefunction`, `QVM`, and `QPU` devices.

* Adds tests for samples

* The tolerance on a couple of other (stochastic) tests was increased, to avoid frequent failures.

**Benefits:**

* A user can train/optimize a hybrid classical-quantum computation, and then, once trained, easily switch the model to generate samples rather than deterministic expectation values/variances.

* This significantly simplifies the code required to compute expectations and variances on hardware. Previously, probabilities had to be computed for multi-qubit observables. Now, we can simply call `np.mean(self.samples())` and `np.var(self.samples())`.

**Drawbacks:** n/a

### CODE REVIEW

**Comment by josh146** [comment]
@msohaibalam, there seems to be an issue with Semaphore - the tests are failing with a Quilc `TimeoutError`?

**Comment by msohaibalam** [comment]
@josh146 This is strange indeed. Let me run it by others and get back to you soon on this. There are a few tests we have in our suite that fail with some probability, but I don't that's what's causing this error.

**Comment by msohaibalam** [comment]
@josh146 All fixed!

File: pennylane_forest/numpy_wavefunction.py [hunk 1]
```diff
@@ -86,14 +82,44 @@ def var(self, observable, wires, par):
...
+        a, P = spectral_decomposition(A)
+
+        p = np.zeros(a.shape)
+
+        for idx, Pi in enumerate(P):
+            p[idx] = self.ev(Pi, wires)
+
+        return np.random.choice(a, n, p=p)
+
     def ev(self, A, wires):
```
Comment (msohaibalam) [inline pennylane_forest/numpy_wavefunction.py]: There's code repetition between this and the same method in `wavefunction.py`. Consider moving into `ForestDevice`.

File: pennylane_forest/wavefunction.py [hunk 2]
```diff
@@ -150,14 +147,44 @@ def var(self, observable, wires, par):
         else:
             A = observable_map[observable]
 
-        var = self.ev(A @ A, wires) - self.ev(A, wires) ** 2
+        if self.shots == 0:
+            # exact variance value
+            var = self.ev(A @ A, wires) - self.ev(A, wires) ** 2
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_forest/wavefunction.py]: Oh that's odd - I've started using Black to autoformat everything to PEP8 standards, which is why  the `**` operators have started getting spaces around them.

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_forest/wavefunction.py]: Looks like this is a known issue, which they are reluctant to fix due to the programmatic complexity:
https://github.com/psf/black/issues/148

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_forest/wavefunction.py]: Fixed.

File: pennylane_forest/wavefunction.py [hunk 3]
```diff
@@ -150,14 +147,44 @@ def var(self, observable, wires, par):
...
+            raise ValueError("The number of samples must be a positive integer.")
+
+        if observable == "Hermitian":
+            A = par[0]
+        else:
+            A = observable_map[observable]
+
+        a, P = spectral_decomposition(A)
+
+        p = np.zeros(a.shape)
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_forest/wavefunction.py]: Done!

File: pennylane_forest/wavefunction.py [hunk 4]
```diff
@@ -46,19 +47,19 @@
 observable_map = {"PauliX": X, "PauliY": Y, "PauliZ": Z, "Identity": I, "Hadamard": H}
 
 
-def spectral_decomposition_qubit(A):
-    r"""Spectral decomposition of a :math:`2\times 2` Hermitian matrix.
+def spectral_decomposition(A):
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane_forest/wavefunction.py]: This is actually a very good call - I made `spectral_decomposition` a static method of the `WavefunctionDevice` class.

At the same time, your comment got me thinking - the `NumpyWavefunctionDevice` and `WavefunctionDevice` are _very_ similar, only differing in the `__init__`, `pre_apply`, and `pre_measure` methods (and this is only due to slightly differing APIs between the PyQVM and QVM).

So what I've done is made `NumpyWavefunctionDevice` inherit from `WavefunctionDevice`. The class is now much simpler, and it has removed a tonne of code duplication!

File: pennylane_forest/numpy_wavefunction.py (hunk 1 above)
Comment (josh146) [reply pennylane_forest/numpy_wavefunction.py]: Addressed as per https://github.com/rigetti/pennylane-forest/pull/18#discussion_r311904647

File: pennylane_forest/wavefunction.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (msohaibalam) [reply pennylane_forest/wavefunction.py]: Awesome! I love removing code :)

**Pull request merged**